### PR TITLE
New Datalog Solver

### PR DESCRIPTION
**Pull request by racerand**
Tasks:
- [ ] Read chapter 3 of the book.
- [ ] Write 20-25 small Datalog benchmarks.
- [ ] Begin to implement the AST for the RAM language. (For, Assign, Loop etc).

### CODE REVIEW

File: main/src/flix/runtime/fixpoint/ram/AssignStmt.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package flix.runtime.fixpoint.ram;
+
+import java.io.PrintStream;
+
+public class AssignStmt implements Stmt {
```
Comment (magnus-madsen) [inline main/src/flix/runtime/fixpoint/ram/AssignStmt.java]: I would make all classes final.

File: main/src/flix/runtime/fixpoint/ram/AssignStmt.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    private TableName name;
+    private RelationExp relationExp;
```
Comment (magnus-madsen) [inline main/src/flix/runtime/fixpoint/ram/AssignStmt.java]: I would make all fields final.

File: main/src/flix/runtime/fixpoint/ram/AssignStmt.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    public AssignStmt(TableName name, RelationExp relationExp) {
+        this.name = name;
```
Comment (magnus-madsen) [inline main/src/flix/runtime/fixpoint/ram/AssignStmt.java]: I suggest you add:

```
if (name == null) throw new IllegalArgumentException("'name' must be non-null");
```

and so forth, and for all constructors.

This is by far the best way to ensure that the RAM ast (which is constructed when the program is running!) remains safe.

File: main/src/flix/runtime/fixpoint/ram/AssignStmt.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        this.relationExp = relationExp;
+    }
+
+    @Override
+    public void prettyPrint(PrintStream stream, int indentLevel) {
```
Comment (magnus-madsen) [inline main/src/flix/runtime/fixpoint/ram/AssignStmt.java]: Looks good 👍

File: main/src/flix/runtime/fixpoint/ram/TableClassifier.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package flix.runtime.fixpoint.ram;
+
+public enum TableClassifier {
+    NEW, // Delta' or new in litterature
+    DELTA,
+    RESULT // the cumulative result
+}
```
Comment (magnus-madsen) [inline main/src/flix/runtime/fixpoint/ram/TableClassifier.java]: I would like some better documentation here :)

File: main/src/flix/runtime/fixpoint/ram/TableClassifier.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package flix.runtime.fixpoint.ram;
+
+public enum TableClassifier {
```
Comment (magnus-madsen) [inline main/src/flix/runtime/fixpoint/ram/TableClassifier.java]: How about "TableVersion" ?

**Comment by racerand** [comment]
All the comments made by @magnus-madsen should have been addresses in 0ea1523 and 04e50c2.

**Comment by magnus-madsen** [comment]
I have been reading through the code. I have some general suggestions for improvement (we can talk about these):

- [ ] Add a LabelStm that can be used to decorate the AST with debugging information. E.g. a string which could be things like "Stratum 1", or "Init" or "Fixpoint" or a reference to the current rule etc.
- [ ] The for loops seems nested in opposite order of the rules. While this is probably still OK, find it hard to read.
- [ ] Why not reuse local variable names? It would be nice if DirectedEdge0 always refers to the first occurence of DirectedEdge in the current rule.
- [ ] Rename `PrintableExp` to `Printable` and have `BoolExp`, `RelationExp` and `Stmt` implement it. Yes I know that expressions don't (currently) use indent level, but so what. Make it uniform :)
 - [ ] Use Intellij ideas reformat on the whole package.
 - [ ] I would consider just deleting `BinaryBoolOperator` and `BinaryBoolExp` and replacing it by two classes `BinaryBoolAndExp` and `BinaryBoolOrExp`
 - [ ] Tuble -> Tuple
 - [ ] Maybe just drop the `BoolExp` and just call them e.g. `AndExp` and `EqualsExp`.
 - [ ] I guess we are mixing the terms "relation" and "table" all over the place. I wonder if we could be consistent.
 - [ ] Why are terms not in their on folder if bool and rel exps are?
 - [ ] `EmptyRelationExp` in the wrong folder?
 - [ ] SeqStmt spurios code in `prettyPrint`. The constructor should ensure that the object is correctly constructed no?
- [ ] `ProjectStm` has a field called "facts", but should this not just be a single fact? In `TubleInRelBoolExp` this is called "terms"
- [ ] `ForEachStm` I guess we also have some confusion between the names "fact", "terms", and "row variable".
- [ ] Use of "toString" and prettyPrint is inconsistent. I would suggest you either implement both everywhere or just pretty print? Possible toString could always call prettyPrint? (e.g. in TableName)

Overall the AST structure looks like its getting really solid (👍 )

**Comment by magnus-madsen** [comment]
I want to review `MySolver` but I think it needs more comments before I have chance, e.g. a method like `findRulesForDerivedInStratum` has no comments, so it is going to be hard for me to read it. I suggest you invest sometime in 50% more documentation and then I will make an attempt at reviewing it.

**Comment by magnus-madsen** [comment]
I think it should also be broken down into a few more smaller functions. That way it will be easier to understand the flow.

**Comment by magnus-madsen** [comment]
@racerand I added source locations to constraints. You should rebase onto master.

(If you have never rebased before, ensure that you have a backup.)

Please also resolve the conflicts in AUTHORS.md with removing anyone.

File: main/src/flix/runtime/fixpoint/ram/stmt/WhileStmt.java [hunk 7]
```diff
@@ -8,6 +8,14 @@
...
 
+    public BoolExp getCondition() {
+        return condition;
+    }
+
+    public Stmt getBody() {
+        return body;
+    }
+
     public WhileStmt(BoolExp condition, Stmt body) {
```
Comment (magnus-madsen) [inline main/src/flix/runtime/fixpoint/ram/stmt/WhileStmt.java]: Constructors should always go before getters/setters per default Java style.

**Pull request closed without merging**